eel/eel-gnome-extensions.c

```diff
@@ -33,78 +33,92 @@
 #include <gtk/gtk.h>
 
 /* Adapted from gio - gdesktopappinfo.c - prepend_terminal_to_vector */
+static const struct {
+    const char *exec;
+    const char *exec_arg;
+    const gboolean escape_command;
+} known_terminals[] = {
+    { "alacritty", "-e", TRUE },
+    { "color-xterm", "-e", TRUE },
+    { "dtterm", "-e", TRUE },
+    { "foot", "--", FALSE },
+    { "gnome-terminal", "--", TRUE },
+    { "kgx", "-e", TRUE },
+    { "kitty", "--", FALSE },
+    { "konsole", "-e", TRUE },
+    { "lxterminal", "-e", TRUE },
+    { "mate-terminal", "-x", TRUE },
+    { "nxterm", "-e", TRUE },
+    { "rxvt", "-e", TRUE },
+    { "sakura", "-x", TRUE },
+    { "st", "-e", TRUE },
+    { "terminator", "-x", TRUE },
+    { "terminology", "-e", TRUE },
+    { "tilix", "-e", TRUE },
+    { "wezterm", "--", TRUE },
+    { "xfce4-terminal", "-x", FALSE },
+    { "xterm", "-e", TRUE },
+};
+
 static char *
 prepend_terminal_to_command_line (const char *command_line)
 {
     GSettings *settings;
-    gchar *prefix = NULL;
+    GString *prefix = NULL;
     gchar *terminal = NULL;
     gchar *ret = NULL;
+    GString *escaped_command_line = NULL;
+    gint i;
 
     g_return_val_if_fail (command_line != NULL, g_strdup (command_line));
 
     settings = g_settings_new ("org.cinnamon.desktop.default-applications.terminal");
     terminal = g_settings_get_string (settings, "exec");
 
     if (terminal != NULL) {
-        gchar *term_path = NULL;
-
-        term_path = g_find_program_in_path (terminal);
-
-        if (term_path != NULL) {
-            gchar *exec_flag = NULL;
+        for (i = 0; i < G_N_ELEMENTS (known_terminals); i++) {
+            if (g_strcmp0 (known_terminals[i].exec, terminal) == 0) {
+                prefix = g_string_new (g_strdup_printf ("%s %s", known_terminals[i].exec, known_terminals[i].exec_arg));
```

Comment:
This should be `g_string_new_take`, since `g_strdup_printf` returns a new string.

```
==928358== 
==928358== 36 bytes in 2 blocks are definitely lost in loss record 14,270 of 37,292
==928358==    at 0x4846828: malloc (vg_replace_malloc.c:442)
==928358==    by 0x5B97937: __vasprintf_internal (vasprintf.c:116)
==928358==    by 0x56A0A32: UnknownInlinedFun (stdio2.h:169)
==928358==    by 0x56A0A32: g_vasprintf (gprintf.c:341)
==928358==    by 0x56715E4: g_strdup_vprintf (gstrfuncs.c:511)
==928358==    by 0x5671940: g_strdup_printf (gstrfuncs.c:541)
==928358==    by 0x234D26: UnknownInlinedFun (eel-gnome-extensions.c:81)
==928358==    by 0x234D26: eel_gnome_open_terminal_on_screen (eel-gnome-extensions.c:138)
==928358==    by 0x217C12: UnknownInlinedFun (nemo-program-choosing.c:201)
==928358==    by 0x217C12: nemo_launch_application_from_command (nemo-program-choosing.c:255)
==928358==    by 0x179AD7: UnknownInlinedFun (nemo-mime-actions.c:1655)
==928358==    by 0x179AD7: activate_callback.lto_priv.0 (nemo-mime-actions.c:1944)
==928358==    by 0x1FAC31: UnknownInlinedFun (nemo-file.c:8764)
==928358==    by 0x1FAC31: file_list_file_ready_callback (nemo-file.c:8754)
```